src/System.Private.CoreLib/shared/System/Buffers/Text/Utf8Parser/Utf8Parser.Number.cs

```diff
@@ -66,34 +72,52 @@ private static bool TryParseNumber(ReadOnlySpan<byte> source, ref NumberBuffer n
 
             if (srcIndex == source.Length)
             {
-                digits[0] = 0;
-                number.Scale = 0;
                 number.IsNegative = false;
                 bytesConsumed = srcIndex;
                 number.CheckConsistency();
                 return true;
             }
 
             int startIndexNonLeadingDigitsBeforeDecimal = srcIndex;
+
             while (srcIndex != source.Length)
             {
                 c = source[srcIndex];
+
                 if ((c - 48u) > 9)
+                {
                     break;
+                }
+
                 srcIndex++;
+                digitCount++;
+
+                if (digitCount >= maxDigitCount)
+                {
+                    // For decimal and binary floating-point numbers, we only
+                    // need to store digits up to maxDigCount. However, we still
+                    // need to keep track of whether any additional digits past
+                    // maxDigCount were non-zero, as that can impact rounding
+                    // for an input that falls evenly between two representable
+                    // results.
+
+                    number.HasNonZeroTail = (c != '0');
+                }
```

Comment:
Why is this change necessary as part of moving this into corelib?  Does this have any impact on throughput, seeing as though we're adding additional logic in what appears to be a few tight loops?

Reply:
CoreFX tests won't pass otherwise.

The Utf8Parser is using the NumberBuffer implementation that was already in S.P.Corelib now. The `Digit` buffer it contains is dynamically sized to be exactly the right size (which is enough digits to hold the largest exact string + 1 digit for rounding and + 1 digit for the null terminator). It also tracks whether any digits that were specified past the rounding digit where zero, which allows it to be correctly rounded in all scenarios.

Reply:
I'll check perf here, but I would expect some minimal impact (most of which should hopefully be handled by the branch predictor).

Reply:
@stephentoub, we have no perf tests that currently exercise this code (which is decimal, double, and float). We only have tests that cover the integer types.
* Working on getting some basic tests to cover this now...

Reply:
Really basic perf tests (just mirroring the utf16 ones) here: https://github.com/dotnet/corefx/pull/33427

Reply:
@stephentoub, @GrabYourPitchforks, perf numbers look similar to the Utf16Parser/Utf16Formatter differences.

For integers (both parsing and formatting), we are the essentially the same (some faster, some slower; all look within std deviation).
For floating-point (decimal, single, and double), we are essentially the same for formatting. For parsing, we essentially match the results that were in the PR changing the UTF16Parser
* For floating-point, we are faster on numbers that are <= 15 digits and with exponents <= 22. For those outside the range, we are ~5x slower but are now returning the correct result.
  * As mentioned on the Utf16Parser issue, we can improve this further for numbers <= 19 digits and with exponents <= 27 by using 80-bit extended precision multiplication/division
    We can improve this for numbers <= 34 digits and with exponents <= 38 by using 128-bit floating-point multiplication/division

Reply:
Can't seem to upload the CSV files right now, GitHub is having issues doing so.